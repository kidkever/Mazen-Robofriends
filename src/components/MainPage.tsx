import React, { Component } from "react";
import CardList from "../components/CardList";
import SearchBox from "../components/SearchBox";
import ErrorBoundary from "../components/ErrorBoundary";
import "./MainPage.css";

interface Props {
  onRequestRobots: any;
  robots: Array<IRobot>;
  searchField: string;
  isPending: boolean;
  onSearchChange: any;
}

interface IRobot {
  name: string;
  id: number;
  email: string;
}

class MainPage extends Component<Props> {
  componentDidMount() {
    this.props.onRequestRobots();
  }

  filterRobots = (robots: Array<IRobot>) => {
    return robots.filter((robot: any) => {
      return robot.name
        .toLowerCase()
        .includes(this.props.searchField.toLowerCase());
    });
  };

  render() {
    const { onSearchChange, robots, isPending } = this.props;
    if (isPending) {
      return <h1 className="tc">Loading</h1>;
    } else {
      return (
        <div className="tc">
          <h1 className="f2 mb0">RoboFriends</h1>
          <SearchBox searchChange={onSearchChange} />
          <ErrorBoundary>
            <CardList robots={this.filterRobots(robots)} />
          </ErrorBoundary>
        </div>
      );
    }
  }
}

export default MainPage;
