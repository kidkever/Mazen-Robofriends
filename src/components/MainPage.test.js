import React from "react";
import { shallow } from "enzyme";
import MainPage from "./MainPage";

let wrapper;

beforeEach(() => {
  const mockProps = {
    onRequestRobots: jest.fn(),
    robots: [
      {
        id: 3,
        name: "john",
        email: "john@gmail.com",
      },
    ],
    searchField: "a",
    isPending: false,
  };
  wrapper = shallow(<MainPage {...mockProps} />);
});

it("renders MainPage without crashing", () => {
  expect(wrapper).toMatchSnapshot();
});

it("filters empty robots", () => {
  expect(wrapper.instance().filterRobots([])).toEqual([]);
});

it("filters robots with wrong search input", () => {
  const mockProps2 = {
    onRequestRobots: jest.fn(),
    robots: [
      {
        id: 3,
        name: "john",
        email: "john@gmail.com",
      },
    ],
    searchField: "a",
    isPending: false,
  };
  const wrapper2 = shallow(<MainPage {...mockProps2} />);
  expect(wrapper2.instance().filterRobots(mockProps2.robots)).toEqual([]);
});

it("filters robots with the right search input", () => {
  const mockProps3 = {
    onRequestRobots: jest.fn(),
    robots: [
      {
        id: 3,
        name: "john",
        email: "john@gmail.com",
      },
    ],
    searchField: "john",
    isPending: false,
  };
  const wrapper3 = shallow(<MainPage {...mockProps3} />);
  expect(wrapper3.instance().filterRobots(mockProps3.robots)).toEqual([
    {
      id: 3,
      name: "john",
      email: "john@gmail.com",
    },
  ]);
});

it("returns an empty object if it is loading / pending", () => {
  const mockProps4 = {
    onRequestRobots: jest.fn(),
    robots: [],
    searchField: "",
    isPending: true,
  };
  const wrapper4 = shallow(<MainPage {...mockProps4} />);
  expect(wrapper4).toEqual({});
});
