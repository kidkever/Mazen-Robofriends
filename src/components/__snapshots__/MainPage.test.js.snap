// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders MainPage without crashing 1`] = `
<div
  className="tc"
>
  <h1
    className="f2 mb0"
  >
    RoboFriends
  </h1>
  <SearchBox />
  <ErrorBoundary>
    <CardList
      robots={Array []}
    />
  </ErrorBoundary>
</div>
`;
